daemon/logger/gelf/gelf.go

```diff
@@ -0,0 +1,114 @@
+package gelf
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/Graylog2/go-gelf/gelf"
+	"github.com/docker/docker/daemon/logger"
+	"github.com/docker/docker/pkg/parsers"
+)
+
+type GelfLogger struct {
+	writer   *gelf.Writer
+	metadata GelfMetadata
+}
+
+type GelfMetadata struct {
+	ContainerId string
+	ImageId     string
+	Entrypoint  string
+	Args        []string
+}
+
+func New(options []string, gelfMetadata GelfMetadata) (logger.Logger, error) {
+	host := "localhost"
+	port := "12201"
+
+	// parse the options
+	for _, option := range options {
+		key, val, err := parsers.ParseKeyValueOpt(option)
+		if err != nil {
+			return nil, err
+		}
+		key = strings.ToLower(key)
+		switch key {
+		case "gelf.host":
+			host = val
+		case "gelf.port":
+			port = val
+		}
+	}
+
+	gelfWriter, err := gelf.NewWriter(fmt.Sprintf("%s:%s", host, port))
+	if err != nil {
+		return nil, fmt.Errorf("Can not connect to GELF endpoint: %v", err)
+	}
+
+	return &GelfLogger{
+		writer:   gelfWriter,
+		metadata: gelfMetadata,
+	}, nil
+}
+
+func (s *GelfLogger) Log(msg *logger.Message) error {
+	// remove trailing and leading whitespace
+	p := bytes.TrimSpace([]byte(msg.Line))
+
+	// If there are newlines in the message, use the first line
+	// for the short message and set the full message to the
+	// original input.  If the input has no newlines, stick the
+	// whole thing in Short.
+	short := p
+	full := []byte("")
+	if i := bytes.IndexRune(p, '\n'); i > 0 {
+		short = p[:i]
+		full = p
+	}
+
+	level := int32(6)
+	if msg.Source == "stderr" {
+		level = 3
+	}
+
+	hostname, err := os.Hostname()
```

Comment:
hostname and command generation is not efficient in `Log` it is better to do them in `New`. Logs very highload part of docker. So it is better to do as less as possible inside `Log`.